### PR TITLE
feat: Storybook7 Upgrade

### PR DESCRIPTION
**Pull request by scotttjob**
## Motivations

Storybook6 is out of date. We needed to move to Storybook7. This gets us there.

## Changes

Upgraded Storybook6 to Storybook7

## Testing

Click on everything inside Storybook and make sure everything is 1:1 with prod.

This is a tedious, manual check. I've gone through things many times myself, but if you're reviewing this I would love a dig around to find any differences. I've tried my best to make things line up 1:1 but in some cases there are minor differences that can't be avoided.

## Old Outstanding Issues (these have been resolved)

1. ~~Our `<details>` sections in our MD files (the README) and some of our `<table>`s are not being parsed/rendered properly in the new MDX2 world with `markdown-to-jsx`. We're using the Storybook provided Markdown component, but it's trying to convert those HTML tags to JSX and it seems to be failing in the process. Need to figure out a path forward here. It might unfortunately be to duplicate the content as moving the `<details>` blocks directly into the mdx file renders perfectly.~~ (I unfortunately had to recode our <details> sections as just normal sections. MDX is moving away from GFM and parsing HTML in general. If you can find another solution to this problem, I am interested in contributions)

2. ~~The image at the top of the README for circleCI is broken. It's a valid MD image link, but markdown-to-jsx doesn't like it. If you import the MD content directly into the mdx file it works fine, but when you pass it through the provided Storybook Markdown component (like with details and table) it doesn't work.~~ (I added a custom renderRule in our MarkdownWrapper. It's ugly, but this is a legit bug in the rendering library so the ugly fix is necessary until they fix it)

3. ~~The MarkdownWrapper component is currently a bunch of differently things cobbled together from .storybook into docs (copying was easier to prove that this was going to work). Before moving this to prod I'd like to clean up that component and see if we can do things in a neater way (import components from .storybook into docs without error)~~ (It's been cleaned up, and context/documentation added but this could still use some more tidy up if anyone is interested)

4. ~~We need to rewrite the docs/stories for InputNumber and InputText. They use `name=` props on their component (see original/prod code) which Storybook is trying to render with their MDX2 changes. For now I've removed the `name=` props so they wouldn't generate an error (and crash the page) but we should find a way to get those back in.~~ (Turns out you just needed to wrap these uses with a <Story> component with a name. That prevents Storybook from trying to render out a story based on the name prop)

5. ~~DataList. We need to rewrite the stories for this one. This component 'renders' in the stories, but it causes the tab to crash which is probably not ideal. I did not debug this one as we should just rewrite the docs in mdx2 format which should alleviate this issue (and we need to do it anyway).~~ (I rewrote the stories. Wasn't too bad. Much faster to load now)

6. ~~There are a few changes in this PR that is just me removing HTML comments in markdown files. I think we just need to wrap those in comments in JS/React style comments `{/** **/}`, but didn't want to spend my timebox time on that problem. The straight HTML comments cause crashing errors.~~ (I tried wrapping in the JS style comments, but our auto-linter then tries to escape them. So I've just removed the commented out blocks)

7. ~~Upgrading to the latest rollup in the design folder (required for this upgrade, errors otherwise) I believe has caused us to generate a bunch of blank css.d.ts files on build for the main/foundation css files which then generate errors on commit/start because they're blank. As a workaround I've put an empty variable in them which makes everything happy but is not a good long-term solution. I would rather uncover the issue under the hood and fix things but I did not have the time to dig in because of my time-box.~~ (I went back and removed the d.ts files and they don't seem to be generating anymore).

---

[In Atlantis we use Github's built in pull request reviews](https://help.github.com/en/articles/about-pull-request-reviews).

Random photo of Atlantis

### CODE REVIEW

**Comment by scotttjob** [comment]
> Hey Scott! It's so exciting to be on Storybook 7 🎉 It's a lot of code to look through so I hope you don't mind if first I note two visual differences I notice.
> 
> I think both are related to `.storybook/preview.tsx`.
> 
>     1. It seems like the `backgrounds` parameter is no longer applying. On the mobile stories, the background colour is no longer `color-surface--background`:
> 
> 
> Storybook 6: Screenshot 2024-01-22 at 11 26 49 AM
> 
> Storybook 7: Screenshot 2024-01-22 at 11 27 00 AM
> 
> Screenshot 2024-01-22 at 11 25 02 AM
> 
>     2. The other thing I noticed is that the `storySort` order is missing some sections:
> 
> 
> Storybook 6: Screenshot 2024-01-22 at 11 29 16 AM
> 
> Storybook 7: Screenshot 2024-01-22 at 11 29 32 AM
> 
> I'll keep digging into why - just wanted to flag these as I mentioned them in Slack the other day and wanted to point them out in the PR early on

Great catches Taylor!

I think I've managed to correct the background issue now, but if you're still seeing it on your end we might have to have a quick pairing session.

Good catch on the missing entries in the nav. As part of the upgrade I had to remove a few paths from our `stories` parameter in main.ts because webpack was insisting on processing all the dist/node_modules directories which preventing anything from building.

I had to manually move some stories around to get them back into the project which is why the modified files has ballooned a bit, but they were all direct moves.

**Comment by scotttjob** [comment]
> Hey Scott, I ran through every component/page and nothing broke in obvious/visible ways. Just the "Code" tab crash I mentioned above, as well as @taylorvnoj's observations 👍

Great catch Jacob!

You've highlighted what might be a larger issue for us though, it looks like that Code tab in general has been moved elsewhere in V7, so we'll need to turn off our manual usages of that code tab and move to the Docs feature like they intend us to to.

https://github.com/storybookjs/storybook/issues/21720

File: packages/design/src/foundation.css [hunk 1]
```diff
@@ -1,11 +1,11 @@
...
+@import "./maincolors.css";
 @import "./radii.css";
-@import "./spacing.css";
+@import "./mainspacing.css";
 @import "./shadows.css";
 @import "./timings.css";
 @import "./opacity.css";
 @import "./elevations.css";
-@import "./typography.css";
+@import "./maintypography.css";
```
[the remark this answers is not in the document]
Comment (scotttjob) [reply packages/design/src/foundation.css:9]: Rollup was upgraded as part of this change, and when doing that it generates some new build artifacts at the root level. Anything with a 'main' added to the name means there are more than 1 file of that name in the project somewhere, and when rollup builds the package it complains about collisions so I had to update a few of the names.

There may be a configuration option that allows us to sidestep this whole problem, but I didn't find it! :)

**Comment by JeffMarvin** [comment]
Noticing a few small things, but nothing that I think should block the release of this. (Especially given the trickiness of managing merge conflicts in the `package-lock.json` file.) 

I think we should release this and then fix small issues in followups. 
1. Typography seems to be different weights, not sure what's going on there.
image

2. Small warnings about some routing 
image

**Review by JeffMarvin — APPROVED** [review]
I added a few small fixes to a couple of issues I found, and disabled the Code tab. As mentioned in the comment, there's a few small things remaining but they are not showstoppers and I recommend we roll forward ever on with much glory and progress! 🚀

**Comment by chris-at-jobber** [comment]
> Typography seems to be different weights

The "Storybook" UI elements in the stories are actually ignoring our custom fonts altogether, it's not just a weights issue. The sidenav and (most importantly) the Atlantis components themselves appear to be correct, so there must be a change in the SB config in how we point to our fonts in the stories themselves. Probably something we can fix in a followup as I know this has gone on for Far Too Long, but hopefully it's trivial as custom fonts in stories is a widely-supported feature of Storybook.

**Comment by taylorvnoj** [comment]
There are some code tabs I've had to hide in the interest of time:
- `docs/components/Chips/Web.stories.tsx` `Selection` story
- `docs/components/Combobox/Web.stories.tsx` `InfiniteScroll` & `CustomSearch` stories
- `docs/components/Page/Web.stories.tsx` for `WithAdditionalTitleFields` which wasn't actually working on master

I also had to remove the `Swatch List` from the `Colors` story. Spoke with Chris about this and this is fine for now, but will try and find a solution in the next few days. All the other color sections are still available to view.

**Review by taylorvnoj — APPROVED** [review]
I did contribute to this PR - but also, most of the team did as well.

All pages have been looked at (Docs, Canvas, Code, etc) in all sections (Components, Design, etc) and nothing is currently broken.

**Pull request merged**